stan/math/gpu/kernel_cl.hpp

```diff
@@ -81,9 +81,9 @@ class kernel_functor {
   kernel_functor(const char* name, const char* source,
                  std::map<const char*, int> options) {
     auto base_opts = opencl_context.base_opts();
-    options.insert(base_opts.begin(), base_opts.end());
-    kernel_ = compile_kernel(name, source, options);
-    opts_ = options;
+    base_opts.insert(options.begin(), options.end());
```

Comment:
The way it was set up before, the default value was never overwritten in cases where the hardware limitations did not allow the default value. 

This meant that on AMD, the 16x16 thread block size was used to spawn the threads, but the kernels were compiled with the THREAD_BLOCK_SIZE 32. 

This switch fixes that. We can look at other solutions if you are not comfortable with this one.

Reply:
it seems roughly equivalent, but I just wanted to double check. When writing the current version I looked up `m.insert()` and saw that it would not overwrite values in `m`. Shouldn't that have the behavior that you're looking for, such that defaults in `base_opts` are not used if they are redefined in `options`?

> Because element keys in a map are unique, the insertion operation checks whether each inserted element has a key equivalent to the one of an element already in the container, and if so, the element is not inserted

(from http://www.cplusplus.com/reference/map/map/insert/)

I would expect this new version to have the behavior that it sounds like you aren't looking for... Unless you are trying to overwrite the `options` with the things in `base_opts` and saying the former are the defaults?

Reply:
I gave this one a second thought. What we actually need is checking whether the `base_opts` value is smaller than the one in `options`, in that case use the one in `base_opts`, otherwise use the options one. The way I changed it we always use the `base_opts` in which case the `options` have no meaning yeah... 

I will rewrite this.

Reply:
Okay I'm pretty sure the above is what we want.

Our default block size is 32, but we also check the whether the sqrt of the `device_thread_block_size` is less than the default. If so than we set the default to the sqrt of the device thread block size. Since that's our max block size we only want to replace it if we have a value that's less. 

@rok-cesnovar do we ever want blocks that are wider than they are long? If so then let's change the default to the full device block size and square blocks can just pass their block size as an option. The only other default options this would effect is `WORK_PER_THREAD` which we can also set as the max block size.